crates/typst/src/eval/mod.rs

```diff
@@ -1692,30 +1692,47 @@ fn apply_imports<V: IntoValue>(
     imports: Option<ast::Imports>,
     vm: &mut Vm,
     source_value: V,
-    name: impl Fn(&V) -> EcoString,
+    new_name: Option<&str>,
+    name: impl FnOnce(&V) -> EcoString,
     scope: impl Fn(&V) -> &Scope,
 ) -> SourceResult<()> {
-    match imports {
-        None => {
-            vm.scopes.top.define(name(&source_value), source_value);
-        }
-        Some(ast::Imports::Wildcard) => {
-            for (var, value) in scope(&source_value).iter() {
-                vm.scopes.top.define(var.clone(), value.clone());
+    if let Some(new_name) = new_name {
+        // Renamed module => just define it on the scope (no items)
+        vm.scopes.top.define(new_name, source_value);
+    } else {
+        match imports {
+            None => {
+                vm.scopes.top.define(name(&source_value), source_value);
             }
-        }
-        Some(ast::Imports::Items(items)) => {
-            let mut errors = vec![];
-            let scope = scope(&source_value);
-            for ident in items.idents() {
-                if let Some(value) = scope.get(&ident) {
-                    vm.define(ident, value.clone());
-                } else {
-                    errors.push(error!(ident.span(), "unresolved import"));
+            Some(ast::Imports::Wildcard) => {
+                for (var, value) in scope(&source_value).iter() {
+                    vm.scopes.top.define(var.clone(), value.clone());
                 }
             }
-            if !errors.is_empty() {
-                return Err(Box::new(errors));
+            Some(ast::Imports::Items(items)) => {
+                let mut errors = vec![];
+                let scope = scope(&source_value);
+                for item in items.items() {
+                    let original_ident = item.original_name();
+                    if let Some(value) = scope.get(&original_ident) {
+                        if let ast::ImportItem::Renamed(renamed_item) = &item {
+                            if renamed_item.original_name().as_str()
+                                == renamed_item.new_name().as_str()
+                            {
+                                vm.vt.tracer.warn(SourceDiagnostic::warning(
+                                    renamed_item.span(),
+                                    "renaming imported name to its own name",
+                                ));
+                            }
+                        }
```

Comment:
I have to admit I'm not particularly proud of having this check here (upon eval) instead of in the parser or AST generator or something like that. Perhaps this could be moved elsewhere, but I'm not sure.

Reply:
same goes for strong and emph emptyness. but since there are no parser warnings atm, there's not much that can be done until we have those.